File: src/main/java/app/tweditor/LoadTemplates.java
package app.tweditor;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Map;
import java.util.Set;
import javax.swing.SwingUtilities;

public class LoadTemplates extends Thread
{
  private final ProgressDialog progressDialog;
  private final AppEnvironment environment;
  private boolean success = false;

  public LoadTemplates(ProgressDialog dialog, AppEnvironment environment)
  {
    this.progressDialog = dialog;
    this.environment = environment;
  }

  public void run()
  {
    try
    {
      Set mapSet = this.environment.getResourceFiles().entrySet();
      int entryCount = mapSet.size();
      this.environment.setItemTemplates(new ArrayList(entryCount));
      int processedCount = 0;
      int currentProgress = 0;

      for (Object mapEntryObj : mapSet) {
        Map.Entry mapEntry = (Map.Entry)mapEntryObj;
        String resourceName = null;
        InputStream in = null;
        Object entryObject = mapEntry.getValue();
        if ((entryObject instanceof File)) {
          File file = (File)entryObject;
          String name = file.getName().toLowerCase();
          int sep = name.lastIndexOf('.');
          if ((sep > 0) && (name.substring(sep).equals(".uti"))) {
            resourceName = name.substring(0, sep);
            in = new FileInputStream(file);
          }
        } else if ((entryObject instanceof KeyEntry)) {
          KeyEntry keyEntry = (KeyEntry)entryObject;
          String name = keyEntry.getFileName().toLowerCase();
          int sep = name.lastIndexOf('.');
          if ((sep > 0) && (name.substring(sep).equals(".uti"))) {
            resourceName = keyEntry.getResourceName();
            in = keyEntry.getInputStream();
          }

        }

        if (in != null) {
          Database database = new Database(this.environment);
          database.load(in);
          in.close();
          DBList fieldList = (DBList)database.getTopLevelStruct().getValue();
          String itemName = fieldList.getString("LocalizedName");
          String itemDescription = fieldList.getString("Description");
          if ((itemName.length() > 0) && (itemDescription.length() > 0)) {
            DBElement resourceElement = new DBElement(11, 0, "TemplateResRef", resourceName);
            fieldList.setElement("TemplateResRef", resourceElement);
            ItemTemplate itemTemplate = new ItemTemplate(fieldList);
            this.environment.getItemTemplates().add(itemTemplate);
          }
        }

        processedCount++;
        int newProgress = processedCount * 100 / entryCount;
        if (newProgress > currentProgress + 9) {
          currentProgress = newProgress;
          this.progressDialog.updateProgress(currentProgress);
        }
      }

      this.success = true;
    } catch (DBException exc) {
      Main.logException("Database error while loading inventory templates", exc);
    } catch (IOException exc) {
      Main.logException("I/O error while loading inventory templates", exc);
    } catch (Throwable exc) {
      Main.logException("Exception while loading inventory templates", exc);
    }

    SwingUtilities.invokeLater(new Runnable() {
      public void run() {
        LoadTemplates.this.progressDialog.closeDialog(LoadTemplates.this.success);
      }
    });
  }
}

File: src/main/kotlin/app/tweditor/LoadTemplates.kt
package app.tweditor

import java.io.File
import java.io.IOException
import java.util.ArrayList

/** Resource-to-template scan used by the Compose shell and headless tests. */
object LoadTemplates {
        /**
         * Parses one resource-map entry into an item template, or null when the
         * entry is not a named .uti. The app runs this on the loader thread;
         * tests call the whole scan directly.
         */
    fun processEntry(entryObject: Any, environment: AppEnvironment): ItemTemplate? {
            val name = ResourceAccess.name(entryObject)?.lowercase() ?: return null
            val sep = name.lastIndexOf('.')
            if (sep <= 0 || name.substring(sep) != ".uti") {
                return null
            }
            val resourceName = when (entryObject) {
                is File -> name.substring(0, sep)
                is KeyEntry -> entryObject.resourceName
                is ResourceEntry -> entryObject.resourceName
                else -> return null
            }
            val input = ResourceAccess.open(entryObject) ?: return null

            val database = Database(environment)
            input.use { database.load(it) }
            val fieldList = database.getTopLevelStruct()!!.getValue() as DBList
            val fullResourceName = name
            val preferredModule = environment.resourceOrigin(fullResourceName)?.moduleName
            val itemName = fieldList.getString("LocalizedName", preferredModule)
            // Description is often empty (books, gems); only a name is required.
            if (itemName.isEmpty()) {
                return null
            }
            val resourceElement = DBElement(11, 0, "TemplateResRef", resourceName)
            fieldList.setElement("TemplateResRef", resourceElement)
            return ItemTemplate(fieldList, preferredModule)
        }

        /** The app's template scan, callable from tests (no progress dialog). */
    fun loadItemTemplates(environment: AppEnvironment) {
            environment.itemTemplates = ArrayList(environment.resourceFiles.size)
            for (mapEntry in environment.resourceFiles.entries) {
                val template = processEntry(mapEntry.value, environment)
                if (template != null) {
                    environment.itemTemplates.add(template)
                }
            }
            environment.icons.primeTemplates(environment.itemTemplates)
    }
}
